apps/omg_watcher/test/omg_watcher/block_validator_test.exs

```diff
@@ -120,5 +118,116 @@ defmodule OMG.WatcherRPC.Web.Validator.BlockValidatorTest do
+      }
+
+      undersize_block = %{
+        hash: "0x0",
+        number: 1000,
+        transactions: []
+      }
+
+      assert {:error, :transactions_exceed_block_limit} = BlockValidator.stateless_validate(oversize_block)
+
+      assert {:error, :empty_block} = BlockValidator.stateless_validate(undersize_block)
+    end
+
+    test "rejects a block that uses the same input in different transactions" do
+      duplicate_input = {1, 0, 0, @alice}
+
+      recovered_tx_1 = TestHelper.create_recovered([duplicate_input], @eth, [{@bob, 10}])
+      recovered_tx_2 = TestHelper.create_recovered([duplicate_input], @eth, [{@bob, 10}])
+
+      signed_txbytes_1 = recovered_tx_1.signed_tx_bytes
+      signed_txbytes_2 = recovered_tx_2.signed_tx_bytes
+
+      block = %{
+        hash: derive_merkle_root([recovered_tx_1, recovered_tx_2]),
+        number: 1000,
+        transactions: [signed_txbytes_1, signed_txbytes_2]
+      }
+
+      assert {:error, :block_duplicate_inputs} == BlockValidator.stateless_validate(block)
+    end
+  end
+
+  describe "stateless_validate/1 (fee validation)" do
+    test "rejects a block if there are multiple fee transactions of the same currency" do
+      input_1 = {1, 0, 0, @alice}
+      input_2 = {2, 0, 0, @alice}
+
+      payment_tx_1 = TestHelper.create_recovered([input_1], @eth, [{@bob, 10}])
+      payment_tx_2 = TestHelper.create_recovered([input_2], @eth, [{@bob, 10}])
+      fee_tx_1 = TestHelper.create_recovered_fee_tx(1, @fee_claimer, @eth, 1)
+      fee_tx_2 = TestHelper.create_recovered_fee_tx(1, @fee_claimer, @eth, 1)
+
+      signed_txbytes = Enum.map([payment_tx_1, payment_tx_2, fee_tx_1, fee_tx_2], fn tx -> tx.signed_tx_bytes end)
+
+      block = %{
+        hash: derive_merkle_root([payment_tx_1, payment_tx_2, fee_tx_1, fee_tx_2]),
+        number: 1000,
+        transactions: signed_txbytes
+      }
+
+      assert {:error, :duplicate_fee_transaction_for_ccy} = BlockValidator.stateless_validate(block)
+    end
+
+    test "rejects a block if fee transactions are not at the tail of the transactions' list (one fee currency)" do
+      input_1 = {1, 0, 0, @alice}
+      input_2 = {2, 0, 0, @alice}
+
+      payment_tx_1 = TestHelper.create_recovered([input_1], @eth, [{@bob, 10}])
+      payment_tx_2 = TestHelper.create_recovered([input_2], @eth, [{@bob, 10}])
+      fee_tx = TestHelper.create_recovered_fee_tx(1, @fee_claimer, @eth, 5)
+
+      invalid_ordered_transactions = [payment_tx_1, fee_tx, payment_tx_2]
+      signed_txbytes = Enum.map(invalid_ordered_transactions, fn tx -> tx.signed_tx_bytes end)
+
+      block = %{
+        hash: derive_merkle_root(invalid_ordered_transactions),
+        number: 1000,
+        transactions: signed_txbytes
+      }
+
+      assert {:error, :unexpected_transaction_type_at_fee_index} = BlockValidator.stateless_validate(block)
+    end
+
+    test "rejects a block if fee transactions are not at the tail of the transactions' list (two fee currencies)" do
```

Comment:
Can we also add a test case in case there is two fee tx of the same output currency?

Reply:
hey Gerhard, what difference do you have in mind from the test on line 160?